packages/noya-state/src/reducers/application.ts

```diff
@@ -249,59 +258,147 @@ export function reducer(
         state.selectedPage = action[1];
       });
     }
-    case 'alignLeft': {
+    case 'distribute': {
       const pageIndex = getCurrentPageIndex(state);
       const layerIndexPaths = getSelectedLayerIndexPaths(state);
+      const axis = action[1];
 
       return produce(state, (state) => {
+        const page = state.sketch.pages[pageIndex];
+        const selectedRect = getSelectedRect(layerIndexPaths, page);
```

Comment:
Let's calculate the `selectedRect` outside of `produce`. 

`getSelectedRect` calls `getBoundingRect`, which traverses every layer

Reply:
I was doing this originally because we write to layer below, I added a comment in the updated code, maybe pulling it out into a util could help document what it's doing better, something like `getWriteableLayer`?

Reply:
I think I see what you mean, how it's not clear whether a value is coming from the `state` within produce or from outside

I'm not sure what the best solution is here, but I agree we should make it more clear. I think we could also propagate the `Draft<T>` type through most functions by using generics, which would help, since then it would be enforced at the type level, to some extent (maybe?)